### ISSUE TITLE
strange behavior of powershell 7.4 

### ISSUE THREAD

**Issue by gssprakash** [PowerShell/PowerShell#20827]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Below command is working fine on powershell 7.4 :
echo Q|openssl.exe s_client -connect www.google.com:443|openssl.exe x509 -noout -dateopt iso_8601 -enddate

But not working with below command with same powershell 7.4:
echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe x509 -noout -dateopt iso_8601 -enddate

error message is :
ResourceUnavailable: Program 'openssl.exe' failed to run: The pipe is being closed.At line:1 char:9
+ echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe …

### Expected behavior

```console
echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe x509 -noout -dateopt iso_8601 -enddate
same above command is working fine with powershell 5.1 as here 

should  get  expiry date of certificate as :
echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe x509 -noout -dateopt iso_8601 -enddate
depth=1 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert SHA2 High Assurance Server CA
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = California, L = Sunnyvale, O = Oath Holdings Inc., CN = *.api.fantasysports.yahoo.com
verify return:1
DONE
notAfter=2023-12-27 23:59:59Z
```

### Actual behavior

```console
error message is :
ResourceUnavailable: Program 'openssl.exe' failed to run: The pipe is being closed.At line:1 char:9
+ echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe …
```

### Error details

```console
error message is :
ResourceUnavailable: Program 'openssl.exe' failed to run: The pipe is being closed.At line:1 char:9
+ echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe …
```

### Environment data

```powershell
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
PS C:\Program Files\PowerShell\7> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

PS C:\Program Files\PowerShell\7> echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe x509 -noout -dateopt iso_8601 -enddate
depth=1 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert SHA2 High Assurance Server CA
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = California, L = Sunnyvale, O = Oath Holdings Inc., CN = *.api.fantasysports.yahoo.com
verify return:1
notAfter=2023-12-27 23:59:59Z
DONE
ResourceUnavailable: Program 'openssl.exe' failed to run: The pipe is being closed.At line:1 char:9
+ echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe …
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.

**Comment by rhubarb-geek-nz** [comment]
I can't reproduce your problem, can you help by..

1. Formatting the exact commands within a code block on the ticket then it is absolutely clear what your commands are.

2. Can you also put spaces around the pipe symbol so it is absolutely clear to both us and the interpreter what are the arguments and which are the pipes

3. Given your example is purely pipes, can you demonstrate the exact same commands running in classic "Command Prompt" ( eg %COMSPEC% )

4. The output of openssl s_client contains much more output that just the x509 certificate, standard process is to [parse the output first before passing to openssl x509](https://www.baeldung.com/linux/ssl-certificates)

5. Can you confirm output with Java's  `keytool -printcert -sslserver mail.yahoo.com:443`

**Comment by gssprakash** [comment]
Installed  powershell 7.4   on machine  having  windows powershell 5.1  and then  executed the below command  and it did not work . But the same command worked fine  on windows powershell 5.1  on the same machine .  

echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe x509 -noout -dateopt iso_8601 -enddate

**Comment by rhubarb-geek-nz** [comment]
I installed 

```
C:> winget install FireDaemon.OpenSSL
Found FireDaemon OpenSSL 3 [FireDaemon.OpenSSL] Version 3.1.4
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://download.firedaemon.com/FireDaemon-OpenSSL/FireDaemon-OpenSSL-x64-3.1.4.exe
  ██████████████████████████████  11.3 MB / 11.3 MB
Successfully verified installer hash
Starting package install...
Successfully installed
```

and confirmed with

```
PS> get-command openssl

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     openssl.exe                                        3.1.4.0    c:\Program Files\FireDaemon OpenSSL 3\bi…
```

The scenario worked on PowerShell 5.1 but not on PowerShell 7.4.0

```
PS C:\Users\rogerb> echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe x509 -noout -dateopt iso_8601 -enddate
depth=1 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert SHA2 High Assurance Server CA
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = California, L = Sunnyvale, O = Oath Holdings Inc., CN = *.api.fantasysports.yahoo.com
verify return:1
notAfter=2023-12-27 23:59:59Z
DONE
ResourceUnavailable: Program 'openssl.exe' failed to run: The pipe is being closed.At line:1 char:9
+ echo Q |openssl.exe s_client -connect mail.yahoo.com:443| openssl.exe …
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
```

I split the pipes up as 

```
PS C:\Users\rogerb> echo Q |openssl.exe s_client -connect mail.yahoo.com:443 > foo.txt
depth=1 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert SHA2 High Assurance Server CA
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = California, L = Sunnyvale, O = Oath Holdings Inc., CN = *.api.fantasysports.yahoo.com
verify return:1
DONE
```

So now have the output data in foo.txt

Then read the foo.txt to get the x509

```
PS C:\Users\rogerb> Get-Content -Raw -AsByteStream -LiteralPath foo.txt | openssl.exe x509 -noout -dateopt iso_8601 -enddate
notAfter=2023-12-27 23:59:59Z
Get-Content: The pipe has been ended.
```

So we have a problem

I then edited the intermediate file to just be the x509 certificate, from the ---BEGIN CERTIFICATE-- to the --END CERTIFICATE---

```
-----BEGIN CERTIFICATE-----
MIIMTjCCCzagAwIBAgIQBv0JpzP3MWq7usgFYBKoSjANBgkqhkiG9w0BAQsFADBw
MQswCQYDVQQGEwJVUzEVMBMGA1UEChMMRGlnaUNlcnQgSW5jMRkwFwYDVQQLExB3
d3cuZGlnaWNlcnQuY29tMS8wLQYDVQQDEyZEaWdpQ2VydCBTSEEyIEhpZ2ggQXNz
...
8NbuD8Op6wu8HwR9/2VUlRPME1xc8gnaON45OhlYas0fZx91WRN/HovNxqHU49Wa
JnnMwU6VixlOAgDPkCIS06Zt6W8oWw3NQ3CBoNWkHC8YhB+8CDhdBPsVcipVP66D
wkEBudQEb4j6mIy9xEKEtXkQjs1T5v/j2O3drZUPQQYffg==
-----END CERTIFICATE-----
```

And then that works cleanly

```
PS> Get-Content -Raw -AsByteStream -LiteralPath foo.txt | openssl.exe x509 -noout -dateopt iso_8601 -enddate notAfter=2023-12-27 23:59:59Z
```

So the problem occurs because when openssl has read the x509 certificate upto the END CERTIFICATE it then ends, because it has done its job

However there is still more data and the write to the pipe fails because the process reading the pipe has finished

Now go back to my point 4, and reference to https://www.baeldung.com/linux/ssl-certificates, it is standard procedure to clean up the output from `openssl c_client` and remove all the garbage that is not part of the certificate before trying to parse it with `openssl x509`

In summary this looks like different behaviour but it is entirely understandable and the error you are seeing is because a component is trying to write more data to a pipe when the reading end has closed. I suggest that this is entirely consistent as an error.

**Comment by rhubarb-geek-nz** [comment]
> verify error:num=20:unable to get local issuer certificate

By that I mean it is entirely consistent as an error to report that it cannot write more data to a pipe because the read has closed, in the same way that it is an error to write to a file when the disk is full.

In this scenario I suggest the correct solution is to clean the certificate before it is read by openssl x509, as Baeldung suggests on Linux...

```
openssl s_client -showcerts -connect baeldung.com:443 </dev/null | sed -n -e '/-.BEGIN/,/-.END/ p' > certifs.pem
```

**Comment by gssprakash** [comment]
what about on windows ?

**Comment by rhubarb-geek-nz** [comment]
> what about on windows ?

Given we are talking about PowerShell here, writing a filter function sounds appropriate to do the equivalent of the the linux sed.

Then you can use the same filter function on all platforms.

**Comment by rhubarb-geek-nz** [comment]
I have one more change to recommend

```
@() | openssl s_client -connect mail.yahoo.com:443 | Filter-X509 | openssl x509 -noout -dateopt iso_8601 -enddate
```

Rather than the [echo Q](url) ( not sure what that is to signify ) the empty array piped into `openssl s_client` results in a stream of zero bytes.

**Comment by jhoneill** [comment]
Is this the  "Pipe no longer adds a trailing break when sending to external executables" thing. ?

This was changed because some things break with the CR/LF 

Does using ```
echo "q`r`n" | ...
```
bring back the behaviour you had on earlier versions ?

**Comment by rhubarb-geek-nz** [comment]
The "echo Q" does nothing except provide a pipe for the input. Providing the empty array does the same thing.

The problem is between the output of openssl s_client and the input of openssl x509 where the s_client will keep on writing more data to the pipe when the x509 reading will stop when it gets to the end of the certificate.

So it is not CR/LF per se, it is that the processes are directly connected by pipes and the write failure is counted as a failure (which it should be, just like a disk-full error).

The filter script I wrote was to parse the output of "openssl s_client" to just write the certificate and not the surrounding "garbage".

**Comment by jhoneill** [comment]
Ah, OK. Ignore what I said then :-)
